### ISSUE TITLE
Prepare for 4.0 release 

### ISSUE THREAD

**Issue by stanfordquan** [gslab-econ/gslab_python#58]
This task will prepare a [changelog](http://keepachangelog.com/en/0.3.0/) for `gslab_python` and `template` 4.0 release. In particular:

1) Document new & removed features, bugs fixed, and changes in existing functionality. Document where they happened. 

2) Document required changes to existing projects' templates due to backward incompatibility.

3) Ping PIs for review.

**Comment by lboxell** [comment]
@stanfordquan,

Some thoughts on this:
 (1) I think our pull-request + squash & merge protocol should take care of this for us. As they will document what was added and when for the major changes (and I'm not sure we want to be documenting the smaller changes). 
 (2) I think the same thing in (1) would apply here too. So, we have some dev-4 template branch, and then that merge into the master branch of template should be the record of what should change for projects using previous templates too. The only exception is features that are backwards incompatible and are not used in template; not sure what the best thing to do in this case is, but perhaps its something we decide as a part of (3) instead of before (3).
 (3) Ping after #59 and we confirm template and gslab_python branches are in-sync?

**Comment by stanfordquan** [comment]
On hold until #59 is done

**Comment by stanfordquan** [comment]
On hold to wrap up Politext

**Comment by lboxell** [comment]
@stanfordquan @arosenbe @M-R-Sullivan @Shun-Yang,

If you don't have other tasks to do, it would probably be good to do no more than an hour of testing of the dev_four branches of gslab_python and template (they should be tested together). That way, we confirm there are no major bugs before MG/JMS review. If you see things that you think could be improved, but aren't bugs, it would probably be worthwhile to note them here too.

**Comment by gentzkow** [comment]
Thanks guys.

A standard we want to hit for the template is that it runs "off the shelf" robustly and with as little setup as possible required. I.e., if we send it to a new user whose machine is not configured to our setup, the setup is simple and clear, and the probability that they get an error the first time they run it is maximized.

I therefore think it would be valuable for you guys to:

1. Set up clean environments w/ no prior setup so you can test the full setup and run process. We should do this for Mac, Linux, and Windows.

2. Devote some effort to thinking about whether there are setup steps we can remove / make optional, ways to improve the instructions in the readme, and anything else that can make the setup more robust.

3.  Spend a little time doing tire kicking to try to find the weak spots that are most likely to produce errors.

Presumably we should make this a separate task to be completed before this task is wrapped up.

**Comment by stanfordquan** [comment]
@lboxell in the original issue description, what I meant is we should have something like [Git's release notes](https://raw.githubusercontent.com/git/git/master/Documentation/RelNotes/2.12.2.txt).

**Comment by lboxell** [comment]
@stanfordquan, That's what I thought you meant. I still think that everything is self-documented if we follow the squash-n-merge protocol (see https://github.com/gslab-econ/gslab_python/compare/master...dev_four). This clearly shows the major changes. If people want more information, its easy to click on a given merge and get more background information.

**Comment by stanfordquan** [comment]
Combined meeting notes:

Meeting note

- README requirements are not currently sequential. 

- Make clearer instruction to install. Put more text into installation instruction, guide users through in a more appealing manner (have examples)

- Echo command for `user-config.yaml`. 

- Allow stata-flavour yaml to be empty. Change stata-flavour to stata-exec. Make comment that Windows user can specify `StataMP-64.exe`.

- Prompt user input for yaml file if it does not exist.

- FAQs (e.g. add PATH)

- `setup.py` script name is not ideal in `template` (`setup.py` means something in the general Python context)

- add link to instructions on how to add environment variables [Mac](https://apple.stackexchange.com/questions/106778/how-do-i-set-environment-variables-on-os-x) [Windows](https://superuser.com/questions/284342/what-are-path-and-other-environment-variables-and-how-can-i-set-or-use-them)

- add `master` to `pip install git+ssh://git@github.com/gslab-econ/gslab_python.git` and other instructions

- drop default support for `google drive` in release

- drop default support for `gslab mtheme`

- git lfs vs cloning: test 

- It's not clear to me why we force the user to have a cache. We should at least discuss that if we make them write it into their `user-config`. (we decided to drop this as default requirement and add it as part of the "extra features" section in the readme.md)

- Remove demonstration of Stata/R, add it to FAQ (`build_stata` vs `build_python`) so forth

- The logging function seems to not be adding newlines in a way that Windows text editors can't read properly (e.g. Notepad)

- The R builder is failing. Command line argument issue (QL fixed in gslab_python)

- `stata yaml`. Maybe we should fork it? Also, add instruction to install:
 
 ```
 net from https://raw.githubusercontent.com/sergiocorreia/stata-misc/master/
 cap ado uninstall yaml
 net install yaml
 ```

**Comment by lboxell** [comment]
@arosenbe @stanfordquan I edited the comment above and gave the task where I thought each item should be completed. Each person can probably close their testing task and open up the new task to implement.

**Comment by lboxell** [comment]
@stanfordquan,

Another item for gslab_python task:

 - Our get_stata_executable() function has the property that if you feed it a valid, non-stata executable it returns that executable. Perhaps we want to add a check that does something like (pseudocode):
```
 if not re.search('stata', os.system('which %s' % stata_exec)):
    return None
```

**Comment by stanfordquan** [comment]
@lboxell,

The original intention of our treatment of Stata executable is that the builder should

1) in the absence of user-specified executable, automatically search for common executables

2) if the user specifies some executable, use it verbatim.

I think we're not doing 1) very well since the yaml file need to be non-empty?

**Comment by stanfordquan** [comment]
This task will now implement the above test comments specific for `gslab_python`. 

- [ ] add master to pip install git+ssh://git@github.com/gslab-econ/gslab_python.git and other instructions [gslab_python task]
- [ ] drop default support for google drive in release [gslab_python task]
- [ ] The R builder is failing in Windows. Command line argument issue. Check if this hold true for others. 
- [ ] Add `setuptools` as a requirement, and suggests `pip` as an installation tool.
- [ ] versioning
- [ ] allow empty yaml?

**Comment by stanfordquan** [comment]
@lboxell, any use for the current stale branches?

**Comment by lboxell** [comment]
@stanfordquan, As long as their noted somewhere on the outline (they contain ideas we hadn't decided whether we wanted to implement yet), you can delete them.

**Comment by lboxell** [comment]
I've found that when there is an error within a script (R, Stata, etc) the error output isn't very informative:
```
Error: Could not call lyx.
              Please check that the executable, source, and target files
              are correctly specified. 
              Command tried: lyx -e pdf2 source/paper/paper.lyx > build/paper/sconscript.log
scons: *** [build/paper/paper.pdf] BadExecutableError : 
Traceback (most recent call last):
  File "/Users/leviboxell/anaconda/lib/python2.7/site-packages/scons-2.5.1/SCons/Action.py", line 1054, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "/Users/leviboxell/anaconda/lib/python2.7/site-packages/GSLab_Tools-3.0.5-py2.7.egg/gslab_scons/builders/build_lyx.py", line 54, in build_lyx
    raise BadExecutableError(message)
BadExecutableError
scons: building terminated because of errors.
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/Users/leviboxell/anaconda/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/Users/leviboxell/anaconda/lib/python2.7/site-packages/GSLab_Tools-3.0.5-py2.7.egg/gslab_scons/log.py", line 45, in end_log
    builder_logs = collect_builder_logs(parent_dir)
  File "/Users/leviboxell/anaconda/lib/python2.7/site-packages/GSLab_Tools-3.0.5-py2.7.egg/gslab_scons/log.py", line 81, in collect_builder_logs
    builder_log_end_time = datetime.strptime(s, "%Y-%m-%d %H:%M:%S")
  File "/Users/leviboxell/anaconda/lib/python2.7/_strptime.py", line 332, in _strptime
    (data_string, format))
ValueError: time data ' restricted \\write18 enabled.' does not match format '%Y-%m-%d %H:%M:%S'
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/Users/leviboxell/anaconda/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/Users/leviboxell/anaconda/lib/python2.7/site-packages/GSLab_Tools-3.0.5-py2.7.egg/gslab_scons/log.py", line 45, in end_log
    builder_logs = collect_builder_logs(parent_dir)
  File "/Users/leviboxell/anaconda/lib/python2.7/site-packages/GSLab_Tools-3.0.5-py2.7.egg/gslab_scons/log.py", line 81, in collect_builder_logs
    builder_log_end_time = datetime.strptime(s, "%Y-%m-%d %H:%M:%S")
  File "/Users/leviboxell/anaconda/lib/python2.7/_strptime.py", line 332, in _strptime
    (data_string, format))
ValueError: time data ' restricted \\write18 enabled.' does not match format '%Y-%m-%d %H:%M:%S'
```

This was printed when lyx couldn't find a figure.

**Comment by stanfordquan** [comment]
There are several things happening here.

The first bit is when you get a `CalledProcessError`. Is it actually because of a CalledProcessError?

The `atexit` error is because the builder didn't finish, so the log-time-reader fails. I'll explore the implications of that bug. I think a try-except block will do fine.

**Comment by lboxell** [comment]
Perhaps we should also point users to the sconscript log for errors in the script?

**Comment by stanfordquan** [comment]
Good idea. I don't know how to separate a `CalledProcessError` into `no executable` vs `process failed` though.

**Comment by stanfordquan** [comment]
Notes while I complete this task

**Comment by stanfordquan** [comment]
When executable calls fail, it can be for any reason, but Python only see a binary of no-error/error. The best we can do is to suggest that either 1) something was not correctly specified or 2) tell user to go look in logs (which we format nicely).

I also realized that it's very cheap to create a Sublime Text syntax that automatically highlights gslab_python-raised errors (since we prefix all such lines with three asterisks `***`), so maybe we can include a link to that in the template README.

**Comment by gentzkow** [comment]
I haven't been following the whole thread, so apologies if this is
redundant, but on the Stata executable issue: Can we dramatically simplify
this by just (i) defining a default Stata executable for each platform;
(ii) allowing the user to specify an alternative in the user settings .yaml
file? We could have the test setup program issue a friendly error message
if Stata is not found suggesting that the user update the executable name
in the settings file. This should allow us to get rid of the need for
environment variables entirely.

(I'd like us to follow the rule of not using environment variables at all
if possible.)
-- 

Matthew Gentzkow
Professor of Economics
Stanford University

**Comment by lboxell** [comment]
@stanfordquan,
 
I don't think switching from 'gslab_tools' --> 'gslab_python' will be a benign switch. I think when you install gslab_tools, it places all the underlying packages inside a gslab_tools distribution folder. Then when you switch the name to 'gslab_python' and install it, you create a gslab_python distribution folder. The issue is that then when you say 'import gslab_scons', it looks through the distribution folders to find gslab_scons and we don't have a guarantee that it takes it from the correct distribution folder. When I tried switching names on my branches, it seemed like python was defaulting to the old distribution (i.e., 'gslab_tools'). I had to pip uninstall gslab_tools before python would import gslab_scons from my new gslab_python distribution. So I vote we leave the name as-is unless we feel like there are non-trivial benefits of switching and we can, with high probability, ensure users don't get confused.

**Comment by stanfordquan** [comment]
On hold.

**Comment by stanfordquan** [comment]
Check @arosenbe 's comment on logging here 
https://github.com/gslab-econ/template/pull/45#issuecomment-304318764

**Comment by stanfordquan** [comment]
Note to self: add an `rclone` builder.

**Comment by arosenbe** [comment]
@stanfordquan, see [here](https://github.com/gslab-econ/gslab_python/issues/52#issuecomment-278149640). We're not sure if we want data download to be part of the scons framework.

**Comment by stanfordquan** [comment]
I reviewed the code and it looks like we have accomplished all that we wanted from the last meeting except for changing `release.py` so that we stop releasing to Google Drive by default.  #68 should be the final task before releasing 4.0

**Comment by stanfordquan** [comment]
Update: pull request for #68 is going on in #72

**Comment by lboxell** [comment]
@stanfordquan,

Can we make figuring out the proper installation instructions so that `python setup.py install clean` deletes any older gslab_tools packages part of this release?

**Comment by stanfordquan** [comment]
We actually implemented everything listed in the checklists from the last meeting! 🥇🥇🥇🥇🥇🥇

PRs [here](https://github.com/gslab-econ/gslab_python/pull/81) and [here](https://github.com/gslab-econ/template/pull/52).
